samply/src/mac/process_launcher.rs

```diff
@@ -5,24 +5,97 @@ use std::io::Write;
 use std::mem;
 use std::os::unix::prelude::OsStrExt;
 use std::path::PathBuf;
-use std::process::{Child, Command};
+use std::process::{Child, Command, ExitStatus};
 use std::sync::Arc;
 use std::time::Duration;
 
 use flate2::write::GzDecoder;
+use mach::task::{task_resume, task_suspend};
+use mach::traps::task_for_pid;
 use tempfile::tempdir;
 
+use crate::shared::ctrl_c::CtrlC;
+
 pub use super::mach_ipc::{mach_port_t, MachError, OsIpcSender};
-use super::mach_ipc::{BlockingMode, OsIpcMultiShotServer, MACH_PORT_NULL};
+use super::mach_ipc::{mach_task_self, BlockingMode, OsIpcMultiShotServer, MACH_PORT_NULL};
+
+pub trait RootTaskRunner {
+    fn run_root_task(&self) -> Result<ExitStatus, MachError>;
+}
 
 pub struct TaskLauncher {
     program: OsString,
     args: Vec<OsString>,
     child_env: Vec<(OsString, OsString)>,
-    _temp_dir: Arc<tempfile::TempDir>,
+    iteration_count: u32,
+}
+
+impl RootTaskRunner for TaskLauncher {
+    fn run_root_task(&self) -> Result<ExitStatus, MachError> {
+        // Ignore Ctrl+C while the subcommand is running. The signal still reaches the process
+        // under observation while we continue to record it. (ctrl+c will send the SIGINT signal
+        // to all processes in the foreground process group).
+        let mut ctrl_c_receiver = CtrlC::observe_oneshot();
+
+        let mut root_child = self.launch_child();
+        let mut exit_status = root_child.wait().expect("couldn't wait for child");
```

Comment:
If samply has debugger entitlements, can we call `task_for_pid` for `root_child.id()` if we're launching system binaries or shell scripts? Or will those still be denied?